### PR TITLE
Harmony 1129 auto pause

### PR DESCRIPTION
**Pull request by indiejames**
## Jira Issue ID
HARMONY-1129

## Description
This PR adds auto-pausing for jobs having more than PREVIEW_THRESHOLD granules. It also activates the `skipPreview` url parameter. When `skipPreview` is `true` there is no auto-pause. 

## Local Test Steps
* Add `PREVIEW_THRESHOLD=10` to your .env file
* Issue a request that will find more than ten granules
* The status page you get redirected to should have a message like
  "The job is generating a preview before auto-pausing. CMR query identified 176 granules, but the request has been limited to process only the first 101 granules because you requested 101 maxResults."
The status should be "previewing"
* Click on the  status page link and refresh the page a few times and you should see the status switch to "paused" with the usual message and 'resume" link
* Either resume or cancel the job to verify it starts running again or switches to the "canceled" status
* Next issue another request with `maxResults` set to 1 and verify that the response is still synchronous.
* Remove the `PREVIEW_THRESHOLD=10` line (or comment it out) in your .env file or you will have many broken tests next time you run `npm run test`

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [ ] ~Documentation updated (if needed)~

### CODE REVIEW

File: app/models/job.ts [hunk 1]
```diff
@@ -87,6 +85,106 @@ export interface JobQuery {
   }
 }
 
+// State machine definition for jobs. This is not used to maintain state, just to enforce
+// transition rules
+const stateMachine = createMachine(
+  {
+    id: 'job',
+    initial: 'accepted',
```
Comment (indiejames) [inline app/models/job.ts:93]: NOTE: We don't ever actually use the 'ACCEPTED' status outside of tests. Jobs are always started as 'RUNNING' or 'PREVIEWING'. We should revisit whether or not we need this anymore.

File: app/models/job.ts [hunk 2]
```diff
@@ -403,10 +506,10 @@ export class Job extends Record implements JobRecord {
   /**
    *  Checks the status of the job to see if the job is paused.
    *
-   * @returns true if the `Job` is paused
+   * @returns true if the `Job` is paused or previewing
    */
   isPaused(): boolean {
-    return this.status === JobStatus.PAUSED;
+    return [JobStatus.PAUSED, JobStatus.PREVIEWING].includes(this.status);
```
Comment (indiejames) [inline app/models/job.ts:512]: I'm not sure if this should be more strict and only be true if the status is PAUSED, but it seemed like it should be considered paused if it were not completed and not running. In any case this function is only used on a test, so maybe not important.

File: env-defaults [hunk 3]
```diff
@@ -104,6 +104,9 @@ MAX_SYNCHRONOUS_GRANULES=1
 # or maximum_async_granules we will override the configuration to limit to this value.
 MAX_GRANULE_LIMIT=2100
 
+# The threshold of the number of granules in a request that will trigger auto-pausing with preview
+PREVIEW_THRESHOLD=500
```
Comment (indiejames) [inline env-defaults:108]: The actual number was not specified in the AC. I can change this to whatever we think the default should be.

**Review by chris-durbin — COMMENTED** [review]
Just added my comments - I'm going to test this out locally now.

**Comment by vinnyinverso** [comment]
When I do skip-preview it seems like the job message doesn't get updated, but the job status does.

```
{"username":"vinverso","status":"running","message":"The job is generating a preview before auto-pausing. CMR query identified 7978 granules, but the request has been limited to process only the first 2001 granules because you requested 2001 maxResults.","progress":0,"createdAt":"2022-05-04T13:53:57.064Z","updatedAt":"2022-05-04T13:54:20.446Z","links":[{"href":"http://localhost:3000/jobs/bc736303-a902-4a24-9ce3-63ec8e13eedd?page=1&limit=2000","title":"The current page","type":"application/json","rel":"self"}],"request":"http://localhost:3000/C1234208438-POCLOUD/ogc-api-coverages/1.0.0/collections/all/coverage/rangeset?maxResults=2001","numInputGranules":2001,"jobID":"bc736303-a902-4a24-9ce3-63ec8e13eedd"}
```

**Review by vinnyinverso — APPROVED** [review]

File: app/models/job.ts [hunk 4]
```diff
@@ -87,6 +86,105 @@ export interface JobQuery {
...
+        id: JobStatus.CANCELED,
+        meta: {
+          defaultMessage: 'The job was canceled',
+        },
+        type: 'final',
+      },
+      previewing: {
+        id: JobStatus.PREVIEWING,
+        meta: {
+          defaultMessage: 'The job is generating a preview before auto-pausing',
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/models/job.ts:149]: I thought about that too, and maybe saying, "click the link below to skip the preview", but then I thought when we add 'cancel' and possibly other links this message might get unmanageable. And the links themselves have their own descriptions.

File: tasks/service-runner/tsconfig.json [hunk 5]
```diff
@@ -4,7 +4,7 @@
       "module": "commonjs",
       "baseUrl": ".",
       "moduleResolution": "node",
-      "target": "ES2018",
+      "target": "ES2019",
```
[the remark this answers is not in the document]
Comment (indiejames) [reply tasks/service-runner/tsconfig.json:7]: Needed to use `Object.fromEntries` in the state machine definition.

File: app/models/job.ts [hunk 6]
```diff
@@ -87,6 +86,105 @@ export interface JobQuery {
...
+          [JobEvent.RESUME, { target: JobStatus.RUNNING }],
+          [JobEvent.CANCEL, { target: JobStatus.CANCELED }],
+          [JobEvent.FAIL, { target: JobStatus.FAILED }],
+        ]),
+      },
+    },
+  },
+);
+
+export const terminalStates = Object.keys(stateMachine.states).filter(key => stateMachine.states[key].type === 'final').map(k => stateMachine.states[k].id) as JobStatus[];
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/models/job.ts]: done

File: app/models/services/base-service.ts [hunk 7]
```diff
@@ -290,16 +290,20 @@ export default abstract class BaseService<ServiceParamType> {
   protected _createJob(
     requestUrl: string,
   ): Job {
+    const url = new URL(requestUrl);
+    const skipPreviewStr = url.searchParams.get('skipPreview');
+    const skipPreview = (this.numInputGranules < env.previewThreshold) || (skipPreviewStr && skipPreviewStr.toLowerCase() === 'true');
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/models/services/base-service.ts]: fixed

File: app/models/services/base-service.ts [hunk 8]
```diff
@@ -290,16 +290,20 @@ export default abstract class BaseService<ServiceParamType> {
...
+    const message = skipPreview ? this.operation.message : [statesToDefaultMessages[JobStatus.PREVIEWING], this.operation.message].join('. ');
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/models/services/base-service.ts]: fixed

**Comment by indiejames** [comment]
Fixed issue with message after clicking skip-preview link

**Review by chris-durbin — APPROVED** [review]
Finished testing out the changes locally - everything is working as expected now.  Nice job with the state machine, the code is more manageable now.

**Pull request merged**